keps/sig-node/753-keystone-containers/README.md

```diff
@@ -0,0 +1,815 @@
+
+The only workaround is that the sidecar container needs some kind of signaling from the app container when it exits or the sidecar monitors the app container status. This is not the ideal workaround as it requires containers to be platform aware.
+
+Another motivation for this KEP is to take small steps towards the sidecar containers that will help a lot of users without including complicated logic and changing critical kubelet code paths.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Pod termination by kubelet if application containers have successfully run to completion and exited.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+- Handling Container startup/shutdown ordering for application and sidecar containers
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+A pod annotation can be specified by users which contains a list of all core application containers in the pod. Depending on the user feedback and how we want to proceed with the sidecar proposal this annotation can be modified or promoted to an API field. 
+
+The annotation is used by kubelet to make a decision on pod termination based on the status of application containers and pod restart policy.
+
+- if all application containers exited successfully and the restart policy of the pod is "never" or "onFailure" then terminate the pod.
+
+- if all app containers are failed and the restart policy is "never" then terminate the pod. 
+If failure logs are concerned in this case (considering the logging container is sidecar) then the sidecar container will only be terminated when all app containers are in the failed state, so it must have captured the failure logs.
+
+The following are existing behaviors just mentioned here for clarity:
+
+- if all application containers exited successfully and the restart policy of the pod is "always" then restart the containers.
+
+- if all app containers are failed and the restart policy is "onfailure", then kubelet will keep trying to restart containers, and no point in terminating the pod
+
+Since container annotations are not supported to be added by an end user, We need to use pod annotation to specify application containers.
+
+There are few options for annotations:
+
+1. Specifying a list of keystone or sidecar containers in one annotation
```

Comment:
Believe this can produce some problems with the usage of mutating webhooks to inject sidecar configurations.

The usage of ``alpha.sidecars.kubernetes.io/<container name>:string` vs `alpha.node.kubernetes.io/keystone:` or `alpha.node.kubernetes.io/sidecars:` reduces cost of uptake of this feature for mutatingwebhook based sidecar injection as the mutators can define their container without requiring the workload owners to uptake the feature.

Reply:
Sorry, not sure I follow what you say. If the list of keystone containers is the value of one annotation, what happens exactly?

Reply:
Definitely agree with the above, as many sidecars are injected the only reliable way to implement this as a webhook implementation would be to add an annotation-per-sidecar container. Anything else has ordering issues in the face of multiple webhooks I think

Reply:
will avoid using `alpha.sidecars.kubernetes.io/<container name>:string` and prefer `alpha.node.kubernetes.io/keystone` or `alpha.node.kubernetes.io/sidecars`

@howardjohn sorry I did not understand single annotation vs list problem

Reply:
Say a user creates a pod with 1 container. There are 2 webhooks injecting 2 sidecars.

* Using the keystone annotation is not viable, because no webhook can safely inject it because they don't know if something is a keystone or not. I would NOT expect users to set that something is a keystone manually when using auto injected sidecars. For example, the kubectl defat container annotation is injected by Istio.
* So that leaves the sidecar annotation. I suppose it's possible to use it but you would need some logic to append to the list if it exists and crate it if not. For example, the gatekeeper example above probably wouldn't work. Note that a real API field probably would since a merge would append to the list.
It also leaves room for misconfigurations of both annotations are set (possibly by 2 different webhooks).

Reply:
Taking into account Tim's input [here](https://github.com/kubernetes/enhancements/pull/2869#issuecomment-999020318), the way to go for now is an annotation for a keystone container, and have an artificial limit of 1 keystone container for now that we can later lift.

This will only be to gather data, we will move to a pod.spec in the future, probably when we implement one of the pre-proposals (we need to decide which one, etc. first)